static/shared/js/typeahead.js.flow

```diff
@@ -0,0 +1,26 @@
+//@flow strict
+
+declare export var popular_emojis: Array<string>;
+
+declare export function remove_diacritics(s: string): string;
+
+declare export function query_matches_source_attrs(
+    query: string,
+    source: {},
+    match_attrs: string,
+    split_char: string,
+): boolean;
+
+declare export function clean_query_lowercase(query: string): string;
+
+declare export function get_emoji_matcher(
+    query: string,
+): (emoji: {}) => boolean;
+
+declare export function triage(
+    query: string,
+    objs: Array<{}>,
+    get_item: (x: {}) => {},
+): {|matches: Array<{}>, rest: Array<{}>|};
```

Comment:
Alright, this is definitely a misunderstanding of the docs on my part.

The functions in this module do look like the accept arbitrary objects, they might be emoji objects but they could be other objects as well.

In the docs here: https://flow.org/en/docs/types/objects/#toc-object-type  they say:

> Sometimes it is useful to write types that accept arbitrary objects, for those you should write {} like this:
> function method(obj: {}) {
>  // ...
> }

So that looked like what I wanted, but I suppose I need to read more carefully, and understand the difference between Object types and "Object" types.

Reply:
Ah, not quite. So "arbitrary objects" would be spelled `{ ... }` (this may be a difference between the latest Flow and the one we're on -- `{}` for us implicitly means `{||}`, while in that bit of the docs it's expected to implicitly mean `{ ... }`.)

But these aren't actually arbitrary. The best way to see that may be to add the types inline: https://github.com/zulip/zulip/pull/21397#issuecomment-1065730079 so you get feedback from the type-checker.

For example, the objects passed to (the function returned by) `get_emoji_matcher` must have a property `emoji_name`, which must be a string.

In the case of this `triage` function, the elements of `objs` are truly arbitrary… *but* there is an important constraint, which is that `get_item` must accept those objects. So the type of the elements of `objs` can be anything, and the type accepted by `get_item` can be anything, but they have to be the *same* anything. That sort of relationship is what [generics](https://flow.org/en/docs/types/generics/) are good for expressing.

Reply:
> (this may be a difference between the latest Flow and the one we're on -- `{}` for us implicitly means `{||}`, while in that bit of the docs it's expected to implicitly mean `{ ... }`.)

Ah, rather, it's the reverse: `{}` implicitly meaning `{ ... }` is the past, and having it implicitly mean `{||}` is the future.

And we've configured Flow to live in that future. From our .flowconfig:
```
; This is the future of Flow; see
;   https://medium.com/flow-type/how-to-upgrade-to-exact-by-default-object-type-syntax-7aa44b4d08ab.
; But we should anticipate that not all our third-party code acts sensibly
; when this is turned on (e.g., see dfa52ab46). As of turning it on, though,
; we're error-free!
exact_by_default=true
```